File: gallery.py
import cv2.aruco as aruco
import cv2
import numpy as np

url = 'https://192.168.0.3:8080/video'
cap = cv2.VideoCapture(url)

aruco_dict = aruco.Dictionary_get(aruco.DICT_6X6_250)
parameters = aruco.DetectorParameters_create()

image = cv2.imread("22.jpg")

while True:
    _, img = cap.read()
    corners, ids, _ = aruco.detectMarkers(img, aruco_dict, parameters=parameters)

    if corners is not None and len(corners) >= 4:

        top_left = None
        top_right = None
        bottom_right = None
        bottom_left = None

        for i, corner in enumerate(corners):
            x = corner[0][0][0]
            y = corner[0][0][1]
            height, width = img.shape[:2]


            if x < width // 2 and y < height // 2:
                top_left = (x, y)
            elif x > width // 2 and y < height // 2:
                top_right = (x, y)
            elif x > width // 2 and y > height // 2:
                bottom_right = (x, y)
            elif x < width // 2 and y > height // 2:
                bottom_left = (x, y)

        if top_left is not None and top_right is not None and bottom_right is not None and bottom_left is not None:
            points = np.array([top_left, top_right, bottom_right, bottom_left], dtype=np.int32)

            pts1 = np.array([top_left, top_right, bottom_right, bottom_left])
            pts2 = np.float32([[0, 0], [width, 0], [width, height], [0, height]])

            resizeimg = cv2.resize(image, (img.shape[1], img.shape[0]))

            matris, _ = cv2.findHomography(pts2, pts1)
            imgOut = cv2.warpPerspective(resizeimg, matris, (img.shape[1], img.shape[0]))
            cv2.fillConvexPoly(img, pts1.astype(int), (0, 0, 0))

            img = cv2.add(imgOut, img)

            cv2.polylines(img, [points], True, (0, 255, 0), 2)

    cv2.imshow("Video", img)
    if cv2.waitKey(1) == ord('q'):
        break

File: gallery.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/aruco.hpp>
#include <iostream>
#include <vector>


using namespace std;
using namespace cv;
int main() {

    aruco::DetectorParameters detectorParams = aruco::DetectorParameters();
    aruco::Dictionary aruco_dict = aruco::getPredefinedDictionary(aruco::DICT_6X6_250);
    aruco::ArucoDetector detector(aruco_dict, detectorParams);

    string url = "https://192.168.0.3:8080/video";
    VideoCapture cap(url);

    Mat image = imread("23.jpg");

    while (true) {
        Mat img;
        cap >> img;


        vector<vector<Point2f>> corners;
        vector<int> ids;
        vector<vector<Point2f>> rejected;
        detector.detectMarkers(img, corners, ids, rejected);

        //aruco::drawDetectedMarkers(img, corners, ids);
        cout<<corners.size();

        if (corners.size() != 0 && corners.size() >= 4)
        {
            Point2f top_left, top_right, bottom_right, bottom_left;
            Size img_size = img.size();

            for (size_t i = 0; i < corners.size(); i++)
            {
                float x = corners[i][0].x;
                float y = corners[i][0].y;
                int width = img.cols;
                int height = img.rows;

                if (x < img_size.width / 2 && y < img_size.height / 2)
                {
                    top_left = Point2f(x, y);
                }
                else if (x > img_size.width / 2 && y < img_size.height / 2)
                {
                    top_right = Point2f(x, y);
                }
                else if (x > img_size.width / 2 && y > img_size.height / 2)
                {
                    bottom_right = Point2f(x, y);
                }
                else if (x < img_size.width / 2 && y > img_size.height / 2)
                {
                    bottom_left = Point2f(x, y);
                }
            }

            if (top_left.x != 0 && top_left.y != 0)
            {

                vector<Point2f> src_points = { top_left, top_right, bottom_right, bottom_left };
                vector<Point2f> dst_points = { Point2f(0, 0), Point2f(img_size.width, 0),Point2f(img_size.width, img_size.height), Point2f(0, img_size.height)};

                Mat resizeImg;
                resize(image, resizeImg, Size(img.cols, img.rows));


                Mat homography = findHomography(dst_points, src_points);
                Mat img_out;

                warpPerspective(resizeImg, img_out, homography, img_size);

                Point polygonPoints[] = { top_left, top_right, bottom_right, bottom_left };
                Scalar_<unsigned char> color(0, 0, 0);

                fillConvexPoly(img, polygonPoints, 4, color);


                img = img_out + img;

                Point pts[] = { top_left, top_right, bottom_right, bottom_left };
                vector<Point> points(pts, pts + sizeof(pts) / sizeof(pts[0]));
                polylines(img, points, true, Scalar(0, 255, 0), 2);

            }
        }

        imshow("Video", img);
        if (waitKey(1) == 'q')
        {
            break;
        }
    }

    return 0;
}
